### PR TITLE
Find integers and floats of equivalent values

### LINKED ISSUE #164: Surprising type comparisons with $signac find
### Description

I have a statepoint parameter that has some `int` values and some `decimal` values. When I use `signac find` to find jobs filtering by this parameter, it only returns jobs where both the value and type of the jobs' statepoints match the search parameters. I would expect the behavior to match that of python's `==` operator, e.g., `$signac find x 10` would return jobs where `job.sp.x = 10` and `job.sp.x = 10.0`.

### To reproduce

```
$ signac init test
Initialized project 'test'.
$ python
>>> import signac
>>> project = signac.get_project()
>>> job = project.open_job({'x': 100**0.5})
>>> job.init()
>>> exit()
$ signac find x 10
Interpreted filter arguments as '{"x": 10}'.
$ signac find x 10.
Interpreted filter arguments as '{"x": 10.0}'.
13dd08001115e65c96a2eeda9ce3505b
```

### System configuration

Please complete the following information:

```
$ python -c 'import platform; print(platform.platform()); import sys; print(sys.version); import signac; print(signac.__version__)'
Darwin-17.7.0-x86_64-i386-64bit
3.5.5 | packaged by conda-forge | (default, Jul 23 2018, 23:45:11) 
[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)]
0.9.5
```

### PR DESCRIPTION
**Pull request by tcmoore3**
Fixes #164

## Description

We find decimal and integer numbers that are mathematically equivalent (e.g., 4 and 4.0) by first checking to see if the value being searched for is an integer (not an `int`, but a "whole" number). If it is, then we do 2 `get`s from the index, one for the `int` value and one for the `float` value. The result in this case is the set union of these two results. 

## Motivation and Context

This issue is described in #164.

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change<sup>1</sup>
- [x] Some middle ground between `Bug fix` and `New feature`

This pull request is based on
- [x] *master*, because it is a bug fix or an update to the documentation.
- [ ] *develop*, because it introduces a new feature.

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.txt).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt).

### CODE REVIEW

File: signac/contrib/collection.py [hunk 1]
```diff
@@ -619,8 +622,25 @@ def _find_expression(self, key, value):
             else:
                 raise KeyError("Unknown expression-operator '{}'.".format(op))
         else:
+
             index = self.index(key, build=True)
-            return index.get(value, set())
+
+            if isinstance(value, Number):
+                is_integer = hash(round(value)) == hash(value)
```
Comment (csadorf) [inline signac/contrib/collection.py]: Would this be equivalent to `round(value) == (value)` ?

File: signac/contrib/collection.py [hunk 2]
```diff
@@ -620,7 +623,19 @@ def _find_expression(self, key, value):
...
+                is_integer = round(value) == value
+            else:
+                is_integer = False
+
+            if is_integer:
+                result_float = index.get(_float(value), set())
+                result_int = index.get(int(value), set())
+                result = result_int.union(result_float)
+            else:
+                result = index.get(value, set())
```
[the remark this answers is not in the document]
Comment (vyasr) [reply signac/contrib/collection.py]: The actual integer line will have to look something like `return index.get(_float(value), set()).union(index.get(int(value), set()))`

[the remark this answers is not in the document]
Comment (csadorf) [reply signac/contrib/collection.py]: Sorry need to object, don't make that a one-liner, please.

[the remark this answers is not in the document]
Comment (vyasr) [reply signac/contrib/collection.py]: So leave the three lines for that bit, just collapse the two conditional blocks?

[the remark this answers is not in the document]
Comment (csadorf) [reply signac/contrib/collection.py]: Yes. Additionally making that one block a one-liner is overkill, because it is harder to read.

**Comment by tcmoore3** [comment]
> I have one change request to simplify the code logic, but otherwise looks got to me overall.
> 
> Did you do some research on whether our `round(value) == value` is best practice?

I did not. The other way to do it is to use the `float.is_integer()` function (introduced in 2.6, see https://stackoverflow.com/questions/21583758/how-to-check-if-a-float-value-is-a-whole-number?answertab=votes#tab-top). I'm not sure which is the best to use in this case.

**Comment by csadorf** [comment]
> > I have one change request to simplify the code logic, but otherwise looks got to me overall.
> > Did you do some research on whether our `round(value) == value` is best practice?
> 
> I did not. The other way to do it is to use the `float.is_integer()` function (introduced in 2.6, see [this discussion](https://stackoverflow.com/questions/21583758/how-to-check-if-a-float-value-is-a-whole-number?answertab=votes#tab-top)). So something like
> 
> ```python
> if isinstance(value, Number) and float(value).is_integer():
>     # do integer stuff
> ```
> 
> I'm not sure which is the best to use in this case.

I like `float.is_integer()`. We can simplify to: `if type(value) is float and value.is_integer():`

We can use the direct type check instead of `isinstance` as a micro optimization, because the filter argument is normalized and any floating-point type is guaranteed to be of type `float` at this point.

**Comment by tcmoore3** [comment]
> I like `float.is_integer()`. We can simplify to: `if type(value) is float and value.is_integer():`

I don't think that will work. We want to find both `int`s and `float`s when we search for either `int`s or integer-valued `float`s, so . E.g., both `signac find x 4` and `signac find x 4.0` should return jobs where `sp.x` is either 4 or 4.0.

**Comment by csadorf** [comment]
> > I like `float.is_integer()`. We can simplify to: `if type(value) is float and value.is_integer():`
> 
> I don't think that will work. We want to find both `int`s and `float`s when we search for either `int`s or integer-valued `float`s. E.g., both `signac find x 4` and `signac find x 4.0` should return jobs where `sp.x` is either 4 or 4.0.

Yeah, you're right. Feel free to leave as is then!

**Comment by tcmoore3** [comment]
> > > I like `float.is_integer()`. We can simplify to: `if type(value) is float and value.is_integer():`
> > 
> > 
> > I don't think that will work. We want to find both `int`s and `float`s when we search for either `int`s or integer-valued `float`s. E.g., both `signac find x 4` and `signac find x 4.0` should return jobs where `sp.x` is either 4 or 4.0.
> 
> Yeah, you're right. Feel free to leave as is then!

I also like `is_integer()` more than `round(x)==x`, so I changed it to use `is_integer`. You may want to double check the changes to the tests, but otherwise this PR should be good to go.

**Pull request merged**